src/cmd/ksh93/tests/vi.exp

```diff
@@ -0,0 +1,670 @@
+# Test for vi keybindings
+
+# vi mode requires being in control mode for ctrl-l to work
+# Send a [ctrl-L] to force the shell to redraw the prompt and command line. Doing so makes it easier
+# to verify the command line contains the expected text.
+proc redraw_prompt args {
+    global is_control_mode
+    if {$is_control_mode == 0} {
+        control_mode
+        send [ctrl L]
+        insert_mode
+    } else {
+        send [ctrl L]
+    }
+#    sleep 1
+}
+
+# Enter vi control mode
+proc control_mode args {
+    global is_control_mode
+    send "\033"
+    set is_control_mode 1
+}
+
+# Enter vi insert mode
+proc insert_mode args {
+    global is_control_mode
+    send "i"
+    set is_control_mode 0
+}
+
+# Delete current line and change input mode to insert mode
+proc clear_prompt args {
+    send [ctrl C]
+    control_mode
+    insert_mode
+}
+
+set pid [spawn $ksh]
+expect_prompt
+
+# Enter vi mode
+send "set -o vi\r"
+expect_prompt
+
+# vi mode starts in insert typing mode
+set is_control_mode 0
+
+# ==========
+# Populate the command history with a couple of commands.
+# Also ensures we are in inset mode
+send "echo hello\r"
+expect "\r\nhello\r\n" {
+    puts "echo 1 works"
+}
+expect_prompt
+send "echo buh bye\r"
+expect "\r\nbuh bye\r\n" {
+    puts "echo 2 works"
+}
+expect_prompt
+
+# ==========
+# [count]k  Fetch previous command.  Each time k is entered the previous command back in time is accessed.
+control_mode
+send "k"
+redraw_prompt
+expect -ex ": echo buh bye" {
+    puts "k fetches previous history item"
+}
+
+control_mode
+send "k"
+redraw_prompt
+expect -ex ": echo hello"
+
+# ==========
+# [count]j  Fetch next command.  Each time j is entered the next command forward in time is accessed.
+control_mode
+send "j"
+redraw_prompt
+expect -ex ": echo buh bye" {
+    puts "j fetches next history item"
+}
+
+# [count]-  Equivalent to k.
+
+# [count][A If cursor is at the end of the line it is equivalent to / with string^set to the contents of the current line.  Otherwise, it is equivalent to k.
```

Comment:
Minor nit but it would be preferable for comment lines to be no longer than 100 chars to match the line length limit for C source.